src/inventory-service/inventory-core/src/main/java/com/inventory/core/InventoryItemService.java

```diff
@@ -71,95 +85,201 @@ public HandlerResponse<InventoryItemDTO> updateStock(UpdateInventoryStockRequest
-            span.setTag("product.currentStockLevel", currentStockLevel);
-            span.setTag("product.newStockLevel", request.getStockLevel());
+            if (span != null) {
+                span.setTag("product.currentStockLevel", currentStockLevel);
+                span.setTag("product.newStockLevel", request.getStockLevel());
+            }
             existingInventoryItem.setCurrentStockLevel(request.getStockLevel());
 
             this.repository.update(existingInventoryItem);
 
-            this.eventPublisher.publishInventoryStockUpdatedEvent(new InventoryStockUpdatedEvent(existingInventoryItem.getProductId(), currentStockLevel, request.getStockLevel()));
+            this.eventPublisher.publishInventoryStockUpdatedEvent(
+                    new InventoryStockUpdatedEvent(existingInventoryItem.getProductId(), 
+                            currentStockLevel, request.getStockLevel()));
 
             return new HandlerResponse<>(new InventoryItemDTO(existingInventoryItem), List.of("OK"), true);
-        } catch (Error error) {
-            logger.error("An exception occurred!", error);
-            span.setTag(Tags.ERROR, true);
-            span.log(Collections.singletonMap(Fields.ERROR_OBJECT, error));
-
-            return new HandlerResponse<>(null, List.of("Unknown error"), false);
+        } catch (Exception e) {
+            logger.error("Error updating stock", e);
+            if (span != null) {
+                span.setTag(Tags.ERROR, true);
+                span.log(Collections.singletonMap(Fields.ERROR_OBJECT, e));
+            }
+            return new HandlerResponse<>(null, List.of("Unknown error: " + e.getMessage()), false);
         }
     }
 
     public HandlerResponse<Boolean> reserveStockForOrder(String orderNumber, List<String> products, String conversationId) {
         final Span span = GlobalTracer.get().activeSpan();
+        logger.info("Reserving stock for order {} with {} products", orderNumber, products.size());
 
-        try {
-            var isFailure = false;
-            ArrayList<InventoryItem> stockAddedFor = new java.util.ArrayList<>();
-
+        if (span != null) {
             span.setTag("order.id", orderNumber);
             span.setTag("order.productCount", products.size());
+            span.setTag("order.conversationId", conversationId);
+        }
 
-            for (var productId : products) {
-                final Span stockCheckSpan = GlobalTracer.get()
-                        .buildSpan("stockCheck")
-                        .asChildOf(span)
-                        .start();
-
-                try (Scope scope = GlobalTracer.get().activateSpan(stockCheckSpan)) {
-                    stockCheckSpan.setTag("product.id", productId);
-                    var inventoryItem = this.repository.withProductId(productId);
+        // Store the products in the order cache early to ensure they're recorded even if reservation fails
```

Comment:
We're in a sync service here I think, right? Do you really need to do this? It feels like it might be one of those "more complexity for more complex failure modes" things

Reply:
Which bit sorry? Using an in-memory cache you mean?

Reply:
yep exactly !